### ISSUE TITLE
Much worse ssd300_coco.py (vgg16) performance as compared to TF's Object Detection API's ssd_mobilenet_v2_coco

### ISSUE THREAD

**Issue by yxchng** [open-mmlab/mmdetection#752]
According to the model zoos of mmdetection and TF's Object Detection API (https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md), performance of `ssd_mobilenet_v2_coco` is mAP `22` and performance of `ssd300_coco.py` is mAP `25.7`, which means `ssd300_coco.py` is much better. Given the stronger backbone, this is expected. However, when I test it, `ssd300_coco.py` is actually much worse.

No matter how I tweak the parameters in the config file, the performance is still worse. It either misses some people or include more than 1 person in a box.

Some examples to illustrate the problem (using default settings for both sides with score_thresholds > 0.5, top is mmdetection's result, bottom is TF's result):

![1_mmdet_vgg](https://user-images.githubusercontent.com/10518587/58871325-5e13aa00-86f4-11e9-9de6-0ec7673267cf.jpg)
![1_tf](https://user-images.githubusercontent.com/10518587/58871331-60760400-86f4-11e9-8862-45b819b10268.jpg)

![2_mmdet_vgg](https://user-images.githubusercontent.com/10518587/58871341-62d85e00-86f4-11e9-9023-97b83ec9868d.jpg)
![2_tf](https://user-images.githubusercontent.com/10518587/58871343-64098b00-86f4-11e9-9ad2-653c0ec29204.jpg)

![3_mmdet_vgg](https://user-images.githubusercontent.com/10518587/58871350-67047b80-86f4-11e9-9bce-6516b48b32bf.jpg)
![3_tf](https://user-images.githubusercontent.com/10518587/58871355-68ce3f00-86f4-11e9-9945-068911266196.jpg)

![4_mmdet_vgg](https://user-images.githubusercontent.com/10518587/58871367-6a980280-86f4-11e9-9000-31ad1f351b6b.jpg)
![4_tf](https://user-images.githubusercontent.com/10518587/58871371-6c61c600-86f4-11e9-9175-64806a9459b8.jpg)

![5_mmdet_vgg](https://user-images.githubusercontent.com/10518587/58871378-6ec42000-86f4-11e9-9bde-1184d8eda1f5.jpg)
![5_tf](https://user-images.githubusercontent.com/10518587/58871381-6ff54d00-86f4-11e9-852c-5d3e8f56ccaa.jpg)

Can I know what is the reason for this phenomena? Is it that the accuracy for `ssd300_coco.py` is wrong? 

I tried training `ssd300_coco.py` with MobileNetV2 backbone but I also cannot get good result. Is the SSD code in this repo not good? Maybe there are bugs?

Hopefully you can help guide me to train a model with good performance. Thanks.

**Comment by yxchng** [comment]
Below is the settings for `ssd_mobilenet_v2_coco`. Seems like the parameters are largely the same. I am posting this here for easy comparison.

```
model {
  ssd {
    num_classes: 90
    image_resizer {
      fixed_shape_resizer {
        height: 300
        width: 300
      }
    }
    feature_extractor {
      type: "ssd_mobilenet_v2"
      depth_multiplier: 1.0
      min_depth: 16
      conv_hyperparams {
        regularizer {
          l2_regularizer {
            weight: 3.99999989895e-05
          }
        }
        initializer {
          truncated_normal_initializer {
            mean: 0.0
            stddev: 0.0299999993294
          }
        }
        activation: RELU_6
        batch_norm {
          decay: 0.999700009823
          center: true
          scale: true
          epsilon: 0.0010000000475
          train: true
        }
      }
      batch_norm_trainable: true
      use_depthwise: true
    }
    box_coder {
      faster_rcnn_box_coder {
        y_scale: 10.0
        x_scale: 10.0
        height_scale: 5.0
        width_scale: 5.0
      }
    }
    matcher {
      argmax_matcher {
        matched_threshold: 0.5
        unmatched_threshold: 0.5
        ignore_thresholds: false
        negatives_lower_than_unmatched: true
        force_match_for_each_row: true
      }
    }
    similarity_calculator {
      iou_similarity {
      }
    }
    box_predictor {
      convolutional_box_predictor {
        conv_hyperparams {
          regularizer {
            l2_regularizer {
              weight: 3.99999989895e-05
            }
          }
          initializer {
            truncated_normal_initializer {
              mean: 0.0
              stddev: 0.0299999993294
            }
          }
          activation: RELU_6
          batch_norm {
            decay: 0.999700009823
            center: true
            scale: true
            epsilon: 0.0010000000475
            train: true
          }
        }
        min_depth: 0
        max_depth: 0
        num_layers_before_predictor: 0
        use_dropout: false
        dropout_keep_probability: 0.800000011921
        kernel_size: 3
        box_code_size: 4
        apply_sigmoid_to_scores: false
      }
    }
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 6
        min_scale: 0.20000000298
        max_scale: 0.949999988079
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.333299994469
      }
    }
    post_processing {
      batch_non_max_suppression {
        score_threshold: 0.300000011921
        iou_threshold: 0.600000023842
        max_detections_per_class: 100
        max_total_detections: 100
      }
      score_converter: SIGMOID
    }
    normalize_loss_by_num_matches: true
    loss {
      localization_loss {
        weighted_smooth_l1 {
        }
      }
      classification_loss {
        weighted_sigmoid {
        }
      }
      hard_example_miner {
        num_hard_examples: 3000
        iou_threshold: 0.990000009537
        loss_type: CLASSIFICATION
        max_negatives_per_positive: 3
        min_negatives_per_image: 3
      }
      classification_weight: 1.0
      localization_weight: 1.0
    }
  }
}
train_config {
  batch_size: 24
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    ssd_random_crop {
    }
  }
  optimizer {
    rms_prop_optimizer {
      learning_rate {
        exponential_decay_learning_rate {
          initial_learning_rate: 0.00400000018999
          decay_steps: 800720
          decay_factor: 0.949999988079
        }
      }
      momentum_optimizer_value: 0.899999976158
      decay: 0.899999976158
      epsilon: 1.0
    }
  }
  fine_tune_checkpoint: "PATH_TO_BE_CONFIGURED/model.ckpt"
  num_steps: 200000
  fine_tune_checkpoint_type: "detection"
}
train_input_reader {
  label_map_path: "PATH_TO_BE_CONFIGURED/mscoco_label_map.pbtxt"
  tf_record_input_reader {
    input_path: "PATH_TO_BE_CONFIGURED/mscoco_train.record"
  }
}
eval_config {
  num_examples: 8000
  max_evals: 10
  use_moving_averages: false
}
eval_input_reader {
  label_map_path: "PATH_TO_BE_CONFIGURED/mscoco_label_map.pbtxt"
  shuffle: false
  num_readers: 1
  tf_record_input_reader {
    input_path: "PATH_TO_BE_CONFIGURED/mscoco_val.record"
  }
}
```

**Comment by hellock** [comment]
The mAP on COCO should be fine.
@yhcao6 You may provide the mAP of vgg_ssd for the person class as a reference.

**Comment by yxchng** [comment]
@hellock  Why the mAP should be fine? These are just some samples of the result. It is worse across many images. I am just curious why this phenomenon happens when it is using a stronger backbone and with the same data.

**Comment by xxllp** [comment]
> @hellock  Why the mAP should be fine? These are just some samples of the result. It is worse across many images. I am just curious why this phenomenon happens when it is using a stronger backbone and with the same data.

how do you predict the image ,could you give me  some  code

**Comment by helloworld77** [comment]
I think ssd is different between mmdetection and TF object detection. The most obvious difference is ssd anchor, mmdetection is same as source ssd, and tf's ssd anchors are more dense than ssd.

**Comment by hellock** [comment]
@helloworld77 If not otherwise specified, the implementation in mmdetection follows original papers for reproduction and fair comparison.

**Comment by helloworld77** [comment]
@hellock Yes, I agree. Actually, the implementation code of ssd anchors is also different between caffe-ssd and mmdetection... although they are all same as the paper. eg. min_size can be multiple in caffe but single in mmdetection, this difference makes it hard to increase anchors in each detection head.
